### DISCUSSION TITLE
Run multiple processes with Dockerfile

### DISCUSSION THREAD

**Discussion by ohld** [dokku/dokku#6299] [General]
Hi team,

When I use Procfile there was an option to run several services within one Dokku app (like `web` instance and `worker`). But now I tend to deploy an app using `Dockerfile` (because deploys tend to be much faster rather than using python buildpacks). 

So there is a question: is there an option to run multiple services within one Dokku app using Dockerfile deployment (like web +worker)?

**Comment by josegonzalez** [comment]
Yeah you can still use a Procfile to set different commands when using a dockerfile.

**Reply by ohld** [reply]
hm, I thought I have to choose between Procfile and Dockerfile deployments. 

Is there any example how can I deploy an app using Dockerfile but then run multiple services using Procfile (web, worker)? I tried to Google but failed. Also I think it may be a great addition to the official documentation.

**Reply by josegonzalez** [reply]
No, you choose between buildpacks and a Dockerfile.

This is all documented in the dockerfile deploy docs: https://dokku.com/docs/deployment/builders/dockerfiles/#procfiles-and-multiple-processes

**Reply by ohld** [reply]
cool. But how is it possible to run two processes (web + worker) while keeping Dockerfile deployment?

**Reply by josegonzalez** [reply]
Literally add a Procfile to your repo. Define your processes in it and then scale them. There isn't anything buildpack-specific about our Procfile support.

**Reply by ohld** [reply]
Gotcha, thanks.

Is there any example how I can run multiple processes defined in Procfile with Dockerfile deployment? I didn't find it in the docs, maybe it is a basic Docker knowledge that I miss.

Imagine that I have Procfile (with `web` and `worker`) and I have scaled them with `app.json` file.  How then should I tweak the Dockerfile to be able to run these several processes? 

E.g. My current Dockerfile ends with this command which runs uvicorn process:
```
CMD ["./scripts/start-dev.sh"]
```

**Reply by polettix** [reply]
My experience with this (I'm still using 0.27 so what I write might not apply any more):

- buildpacks-based workflows
    -  produce images that have no `Entrypoint` at all (it's `null` in the output of `docker inspect ...`)
    - when Dokku starts them, it passes a `Cmd` like `['/start', 'web']`, as seen from the output of `dokku ps:inspect my-app`
- Dockerfile-based workflows (including when using a pre-build image acquired with `docker load`, which is wrapped in Dockerfile)
    - if it has an `Entrypoint` set, Dokku sets no `Cmd`. In my understanding, this disables support for `Procfile` completely.

My guess, and my next attempt, will be to remove the `Entrypoint` frome the image and see if the `['/start', '/web`]` command re-appears.

**Reply by polettix** [reply]
Additional findings:

- creating an image with an empty `ENTRYPOINT` seems to have done the trick. Keep in mind that whatever command is executed from the image's **working directory**, so either use absolute paths or make sure to set it to the right place (in my case it was `/app`, not sure now if it's something Dokku insists on)
- the `/start <whatever>` is specific to buildpacks. The command provided to `docker` was an expansion from the `Procfile` inside the image (in `/app/Procfile`), which is remarkable as that file lives *inside* the container image and the expansion is done outside.

**The difference is important IMHO**, because there are two different pieces of code doing the parsing/expansion of variables in the two cases, so bugs might affect them differently.

Other than that, to use the `Procfile` approach I'd suggest the following:
- make sure to set an empty `ENTRYPOINT`
- make sure to set the `WORKDIR` to the directory where the `Procfile` lives

**Reply by josegonzalez** [reply]
I'm pretty sure this is documented but:

- The `/start PROCESS_TYPE` command is exclusive to Herokuish images. We let herokuish do the right thing for itself.
- Any files are extracted from the repository source, as pushed to Dokku. If your app is deployed via Docker image, we'll extract the file from the `WORKDIR` of the image (with fallbacks if no such path is set).
- A `Procfile` is required for a buildpack-based app unless the buildpack injects a default Procfile itself (some do).
- For all other images (whether you build via dockerfile, pack, deploy an image directly, or whatever), we pass the command specified in your process type's `Procfile` as the command. This overrides any `CMD` specified.
- If not using herokuish, we expand the command from the `Procfile` on the host using app env vars, then pass that in as the command.
- If your docker image/dockerfile has an `ENTRYPOINT`, any command is passed to that.
- If your app has no `Procfile`, we'll just run the container as is. If it doesn't specify a CMD or ENTRYPOINT, it might fail.
- Commands generally run in the `WORKDIR` of your image, unless you are missing one, in which case we just do whatever docker does.

This is my recollection of how the behavior should be in the latest Dokku. If that doesn't seem to be the case, thats a bug.
